### PR TITLE
Dev mode - make it possible to watch files matching a predicate

### PR DESCRIPTION
**Pull request by mkouba**
- HotDeploymentWatchedFileBuildItem#builder().setPredicate(p).build()

See also https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Allow.20extensions.20to.20watch.20new.20files.20.28with.20restart.29

### CODE REVIEW

File: core/devmode-spi/src/main/java/io/quarkus/dev/testing/TestWatchedFiles.java [hunk 1]
```diff
@@ -11,19 +14,23 @@
 public class TestWatchedFiles {
 
     private static volatile Map<String, Boolean> watchedFilePaths;
-    private static volatile Consumer<Map<String, Boolean>> watchedFilesListener;
+    private static volatile BiConsumer<Map<String, Boolean>, List<Entry<Predicate<String>, Boolean>>> watchedFilesListener;
+    private static volatile List<Entry<Predicate<String>, Boolean>> watchedFilePredicates;
 
-    public synchronized static void setWatchedFilePaths(Map<String, Boolean> watchedFilePaths) {
+    public synchronized static void setWatchedFilePaths(Map<String, Boolean> watchedFilePaths,
```
Comment (mkouba) [inline core/devmode-spi/src/main/java/io/quarkus/dev/testing/TestWatchedFiles.java:20]: I'm not so sure about these changes. It's a breaking change in the `devmode-spi`. And I have no idea if there's something out there using this SPI.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/HotDeploymentWatchedFileBuildItem.java [hunk 2]
```diff
@@ -32,4 +73,37 @@ public boolean isRestartNeeded() {
...
 
+    public static class Builder {
+
+        private String location;
+        private Predicate<String> predicate;
+        private boolean restartNeeded = true;
+
+        public Builder setLocation(String location) {
+            if (predicate != null) {
+                throw new IllegalArgumentException("Location already set");
```
Comment (Ladicek) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/HotDeploymentWatchedFileBuildItem.java:84]:
```suggestion
                throw new IllegalArgumentException("Predicate already set");
```
?

Comment (mkouba) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/HotDeploymentWatchedFileBuildItem.java:84]: Oh, sure :facepalm:

File: core/devmode-spi/src/main/java/io/quarkus/dev/testing/TestWatchedFiles.java (hunk 1 above)
Comment (Ladicek) [reply core/devmode-spi/src/main/java/io/quarkus/dev/testing/TestWatchedFiles.java:20]:
```java
@Deprecated(forRemoval = true)
public synchronized static void setWatchedFilePaths(Map<String, Boolean> watchedFilePaths) {
    setWatchedFilePaths(watchedFilePaths, List.of());
}
```

should preserve compatibility pretty easily.

Comment (mkouba) [reply core/devmode-spi/src/main/java/io/quarkus/dev/testing/TestWatchedFiles.java:20]: Of course, but I'm not sure it's worth the effort. I don't even grasp the meaning of `TestWatchedFiles`  :shrug:

Comment (mkouba) [reply core/devmode-spi/src/main/java/io/quarkus/dev/testing/TestWatchedFiles.java:20]: But you're probaly right. Let's deprecate the old methods...

**Comment by Sanne** [comment]
I haven't checked how this works, but at high level I wonder how this can possibly work. We don't watch every single file on the system, do we? So how can you possibly filter among the whole universe?

I suspect the term "predicate" is used in a different way than I'd expect, it might be worth clarifying at least javadoc, if not finding a better name.

**Comment by ia3andy** [comment]
> I haven't checked how this works, but at high level I wonder how this can possibly work. We don't watch every single file on the system, do we? So how can you possibly filter among the whole universe?
> 
> I suspect the term "predicate" is used in a different way than I'd expect, it might be worth clarifying at least javadoc, if not finding a better name.

maybe naming it locationPredicate as the predicate is on the location?

**Comment by mkouba** [comment]
> I haven't checked how this works, but at high level I wonder how this can possibly work. We don't watch every single file on the system, do we? So how can you possibly filter among the whole universe?

@Sanne Of course we don't filter the whole universe. We do filter the set of detected changes. This set is not well documented but from the code it looks like we watch resource dirs of all reloadable modules (as defined in `DependenciesFilter.getReloadableModules(ApplicationModel)`).

> 
> I suspect the term "predicate" is used in a different way than I'd expect, it might be worth clarifying at least javadoc, if not finding a better name.

I will try to add something to the javadoc. Out of curiosity, what's the way you would expect?

**Comment by mkouba** [comment]
> > I haven't checked how this works, but at high level I wonder how this can possibly work. We don't watch every single file on the system, do we? So how can you possibly filter among the whole universe?
> > I suspect the term "predicate" is used in a different way than I'd expect, it might be worth clarifying at least javadoc, if not finding a better name.
> 
> maybe naming it locationPredicate as the predicate is on the location?

Makes sense. I will also add some more info to the javadoc.

**Comment by mkouba** [comment]
Those JVM Tests - JDK 19 failures do not seem to be related...

**Comment by ia3andy** [comment]
I will give it a try today, thanks @mkouba

**Comment by mkouba** [comment]
> I will give it a try today, thanks @mkouba

Hi Andy, did you manage to try this out?

**Review by ia3andy — APPROVED** [review]
It's perfect @mkouba thanks again and sorry for the delay

**Comment by mkouba** [comment]
@gsmet I cannot make the JVM Tests - JDK 19 pass :-(. But the various failures do not look related...

**Comment by ia3andy** [comment]
Awesome I was waiting for this one :) thanks agian @mkouba

**Pull request merged**